api/rosetta/payloads.go

```diff
@@ -0,0 +1,128 @@
+	Transaction string                  `json:"unsigned_transaction"`
+	Payloads    []object.SigningPayload `json:"payloads"`
+}
+
+// Payloads implements the /construction/payloads endpoint of the Rosetta Construction API.
+// It receives an array of operations and all other relevant information required to construct
+// an unsigned transaction. Operations must deterministically describe the intent of the
+// transaction. Besides the unsigned transaction text, this endpoint also returns the list
+// of payloads that should be signed.
+// See https://www.rosetta-api.org/docs/ConstructionApi.html#constructionpayloads
+func (c *Construction) Payloads(ctx echo.Context) error {
+
+	var req PayloadsRequest
+	err := ctx.Bind(&req)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidEncoding(invalidJSON, err))
+	}
+
+	if req.NetworkID.Blockchain == "" {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(blockchainEmpty))
+	}
+	if req.NetworkID.Network == "" {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(networkEmpty))
+	}
+
+	// Metadata object is the response from our metadata endpoint. Thus, the object
+	// should be okay, but let's validate it anyway.
+	rosBlockID := req.Metadata.CurrentBlockID
+	if rosBlockID.Index == nil && rosBlockID.Hash == "" {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(blockEmpty))
+	}
+	if rosBlockID.Hash != "" && len(rosBlockID.Hash) != hexIDSize {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(blockLength,
+			withDetail("have_length", len(rosBlockID.Hash)),
+			withDetail("want_length", hexIDSize),
+		))
+	}
+
+	intent, err := c.transact.DeriveIntent(req.Operations)
+	var iaErr failure.InvalidAccount
+	if errors.As(err, &iaErr) {
+		return echo.NewHTTPError(http.StatusUnprocessableEntity, invalidAccount(iaErr))
+	}
+	var icErr failure.InvalidCurrency
+	if errors.As(err, &icErr) {
+		return echo.NewHTTPError(http.StatusUnprocessableEntity, invalidCurrency(icErr))
+	}
+	var ucErr failure.UnknownCurrency
+	if errors.As(err, &ucErr) {
+		return echo.NewHTTPError(http.StatusUnprocessableEntity, unknownCurrency(ucErr))
+	}
+	var opErr failure.InvalidOperations
+	if errors.As(err, &opErr) {
+		return echo.NewHTTPError(http.StatusUnprocessableEntity, invalidFormat(txInvalidOps))
+	}
+	var inErr failure.InvalidIntent
+	if errors.As(err, &inErr) {
+		return echo.NewHTTPError(http.StatusUnprocessableEntity, invalidIntent(inErr))
+	}
+	if err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, internal(intentDetermination, err))
+	}
+
+	unsigned, err := c.transact.CompileTransaction(req.Metadata.CurrentBlockID, intent, req.Metadata.SequenceNumber)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError, internal(txConstruction, err))
+	}
+
+	sender := identifier.Account{
+		Address: intent.From.String(),
+	}
+	hash, err := c.transact.HashPayload(req.Metadata.CurrentBlockID, unsigned, sender)
```

Comment:
I don't like that the current block ID is part of this interface to be honest.
It implies that the current block plays a role in the transaction hash, which it doesn't.

I realize that it's due to the current structure, but a more accurate formulation for me would be something like `c.transact.HashPayload(unsigned, hashAlgo)` where the input is only the hash algorithm.

Also, transactions that reference a block can be valid for X blocks.
If the reference block is block with height 10, and the latest block is block 14, we want to get the key at block 14, regardless of when the transaction started forming.

Reply:
@Maelkum Could you please elaborate on this in an issue? I missed the comment because it was outdated already.